### ISSUE TITLE
Large response bodies cause hanging

### ISSUE THREAD

**Issue by shockey** [swagger-api/swagger-ui#3832]
| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  Bug
| Which Swagger/OpenAPI version?  | 2.0
| Which Swagger-UI version?       | 3.4.0
| How did you install Swagger-UI? | local dev server
| Which browser & version?        | Chrome 61
| Which operating system?         | macOS High Sierra

### Demonstration API definition

Swagger 2 Petstore is sufficient.

### Configuration (browser query string, constructor, config.yaml)

Default.

### Current Behavior

As an example, `GET /pet/findByStatus` in the Petstore currently returns 1.8MB of JSON when all statuses are selected. This is a pretty big chunk of data, but Swagger-UI insists on rendering the _entire_ body, which is neither performant or useful. The main thread of the application is locked for upwards of 20 seconds on my machine.

### Possible Solution

Truncate or refuse to display large textual response bodies.

### Context

`GET /pet/findByStatus` is my favorite operation to use when testing UI enums, so this is something that impedes me pretty regularly.

**Comment by shockey** [comment]
cc @webron, would appreciate a priority assignment

**Comment by heldersepu** [comment]
This is strange!
I swear I've used larger responses without any issues, let me see if I can make one up...

**Comment by heldersepu** [comment]
Here is my test:
http://swashbuckletest.azurewebsites.net/swagger/ui/index?filter=Huge#/HugeResponse/HugeResponse_Get
Yes we have a problem even with 100000 it generates a 378K response that takes way to long for my liking

**Comment by heldersepu** [comment]
I think that a quick fix will be not to display such a large responses instead show something like github does on diff:
Large diffs are not rendered by default.

**Comment by heldersepu** [comment]
[Profile-20171028T214335.zip](https://github.com/swagger-api/swagger-ui/files/1424739/Profile-20171028T214335.zip)
Attached is my performance profile I hope that can help troubleshooting.
something in react is taking too long
webpack:///./~/react-dom/lib/DOMChildrenOperations.js?568f
```js
function removeChild(parentNode, childNode) {
  if (Array.isArray(childNode)) {
    var closingComment = childNode[1];
    childNode = childNode[0];
    removeDelimitedText(parentNode, childNode, closingComment);
    parentNode.removeChild(closingComment);
  }
  parentNode.removeChild(childNode);
}
```

**Comment by webron** [comment]
As @heldersepu suggested, I believe it's fair to limit the size that's being rendered, and if it's larger, make a link to download the result separately. This is not perfect though, as it means running the call again, and for some APIs it won't necessarily return the same result.

**Comment by shockey** [comment]
@webron, we'd be able to store the response string we already received, and allow the user to download it as a file - no second request needed.

As a proof of concept, see the `Save as JSON/Save as YAML` options in Swagger-Editor: we take the current editor content as a string and download it as a file, all from the client side.

**Comment by webron** [comment]
That's perfect then.

**Comment by heldersepu** [comment]
@shockey & @webron Love the solution! there download link will be awesome!

...but someone should try looking into why the slowdown
You guys don't know any react guru that can truly get to the bottom of the issue?

**Comment by webron** [comment]
In the past, we had issues with the syntax highlighter and large responses, but that doesn't seem to be the issue. I've tested the same API call in both Nightly and Chrome, and they both render fairly quickly (~3 seconds) when the format is JSON. When the format is XML though... that's when it explodes.

**Comment by shockey** [comment]
Okay, I dug into this a bit more, and @webron's note about XML is spot on.

When rendering XML, React spends a _ton_ of time in the `ResponseBody` render function:

<img width="1113" alt="messages image 4167817450" src="https://user-images.githubusercontent.com/680248/32297623-efaebe0c-bf0c-11e7-96b5-d4d17fcc9004.png">

Everything else looks performant, so the good news is that our problem is contained to one component.

Doing a traditional profile on call stack times between clicking Execute and seeing the body revealed the true issue:

<img width="1306" alt="messages image 880229148" src="https://user-images.githubusercontent.com/680248/32297775-79f43628-bf0d-11e7-9d1d-8d0a5b5b900a.png">

More than 95% of the main thread's time is spent parsing the whitespace-stripped XML that comes from the server into a pretty™ XML string.

### TL;DR: it's the `formatXml(xmlStr)` prettifier function in `utils.js`

**Comment by heldersepu** [comment]
I was looking at the formatXml function and that code smells funny...
Inside that function there is a section where we:
- create an array `types` 
- loop over types populating `results` 
- finally get the first one of result

It might be late and I'm missing something ...
...or that can be simplified, I'm sending you guys a PR

**Comment by shockey** [comment]
@heldersepu, I agree - note that the code was originally taken from an SO answer.

I'd be in favor of using a third-party module to handle the formatting. In the meantime, I'll take a look at your PR 😄

**Comment by heldersepu** [comment]
I think we still have a problem with large responses, my test:
http://swashbuckletest.azurewebsites.net/swagger/ui/index?filter=Huge#/HugeResponse/HugeResponse_Get
that is a json response and it gets uncomfortably slow to render...

I think we should still set a hard MAX for the response size  and provide the download link for those large ones

**Comment by shockey** [comment]
@heldersepu, yeah, I see it with your example - the request comes back in under a second, but the rendering takes about ten seconds. I agree with the hard max, but I'd also like to continue looking at the underlying causes and try to address as we go.

Reopening.

Looks like the lion's share of that time (about 8 seconds) is React fiddling with the DOM. The response rendering causes 370(!!) DOM operations to be queued, and >90% of them appear to be caused by JumpToPath affecting ModelCollapse, which is unrelated to the response coming in as far as I can see.

This is a performance issue, but I'm inclined to keep this separated from our big performance ticket, since it's related to Try-It-Out, not initial rendering.

Below is a snapshot of React's DOM operation queue (the HighlightCode update with the response body isn't even on the first page):

<img src='https://shipusercontent.com/231eea278039c4e2184d8a10d24d81aa/Pasted%20Image%201.png' title='Pasted Image 1.png' width=1281>

**Comment by masc3d** [comment]
why not render asynchronously in stages, like eg. postman does. 
it works well and there's barely any need for switches or limits.

**Comment by masc3d** [comment]
btw this issue got much worse in recent versions, presumably due to pretty-print / color enhancements which make rendering even more expensive.

**Comment by heldersepu** [comment]
> why not render asynchronously in stages, like eg. postman does.
> it works well and there's barely any need for switches or limits.

What's the purpose of showing/render a response larger than 2MB?
That is certainly not humanly readable, my sample above on postman takes a few seconds too on the "formating" step...
To me what github does on diff is the way to go,  let the users know the response is large and give them an option to download content or render at own risk

**Comment by masc3d** [comment]
> What's the purpose of showing/render a response larger than 2MB?

your argument makes no sense to me.
if there's no purpose (in showing the content), why would you download it to eventually do exactly that.

download is okay, it would resolve the most important aspect of this issue. 
but it's still less convenient.

**Comment by heldersepu** [comment]
here are a couple of ideas why someone would download:
- to send it to someone else
- to processes it with a tool 
Hopefully, that makes the argument make sense to you

**Comment by masc3d** [comment]
if it makes sense to download it (very often) makes even more sense to view the content right in place.

just for quick checking a subset of a larger result, having to download, open in another editor and cleaning up afterwards are just and awful lot of unnecessary and tiring additional steps.

**Comment by heldersepu** [comment]
>if it makes sense to download it (very often) makes even more sense to view the content right in place.
YES, and that is exactly what Swagger-UI is doing right now, it works perfectly for 99.9% of the cases ...
but this issue is about those not very often cases, like those with large response bodies 

There are purpose-built tools, Swagger-UI is no exception...
From my days creating ETL jobs, I would not use a browser or a regular text editor to look at large data, instead, I get the schema definition, and load the data into a database for further processing/analysis.

This issue is more around bad practices on API design, responses should not be large, and if there is a need the architect should use pagination, that is why personally never attempted a code fix.

**Comment by masc3d** [comment]
> This issue is more around bad practices on API design, responses should not be large,
highly depends on the use case.

anyway it's not implemented that way in postman for no reason,
expensive operations should not block UI, talking about bad practices.

in latest builds it doesn't even take a really "large" result for swagger-ui to begin stalling.

**Comment by heldersepu** [comment]
>anyway it's not implemented that way in postman for no reason, expensive operations should not block UI, talking about bad practices.

Remember the project is open source, if this is a big issue to you the code is there for you to fix it

**Comment by masc3d** [comment]
> Remember the project is open source, if this is a big issue to you the code is there for you to fix it

it's not an issue at all for me.
just some thoughts for a good solution, as it wasn't mentioned as an option before.

**Comment by jmsteinbrunner** [comment]
### Q&A (please complete the following information)
 - OS: Windows 10.0.18363 Build 18363
 - Browser: Chrome 85.0.4183.102
 - Method of installation: Swashbuckle.AspNetCore 5.6.0 NuGet
 - Swagger-UI version: 3.32.5

I'd like to tag on to this issue.  I've been using Swashbuckle.AspNetCore 5.5.1 (Swagger-UI v3.26.0 I believe) and have a somewhat large response body (20k lines of json give or take).  In the 5.5.1, the request duration takes 45ms and displays almost instantly.  Using the exact same configuration with 5.6.0 (Swagger-UI v.3.32.5), request duration takes 45ms and locks the interface 30 seconds and finally displays.  Once that is rendered, if I minimize or expand anything else, the interface is locked again for 30 seconds each time.  Once I clear the results, the interface speed returns to normal.

The coloring added is fantastic but I am currently stuck using the older version as the speed is simply too much of an issue.

**Comment by Omzig** [comment]
In order for the results to be displayed i have to .Take(200), if i let more than go through, it will not render and times out. When i revert to 5.5.1 the UI will render all the records almost instantly. I am now stuck with Swashbuckle.AspNetCore 5.5.1

**Comment by lsanders** [comment]
Ran into this recently too. Looks like you can turn it off via config. For me, this worked:

```
SwaggerUI({
        syntaxHighlight: {
          activated: false,
          theme: "agate"
        },
        //url: path,
        ....
      });
```

**Comment by a-v-e-s** [comment]
Is there any reason some response bodies would try to display in spite of the limit set via Tools ==> Options ==> Display ==> Large Request/Response Min Size ? I have both set to 100,000 which I believe is the default. Normally large responses are not rendered, but sometimes Zap decides to try anyways and I just had a 3.6 MB response freeze my session for about 20 minutes before I manually killed the process.

**Comment by EngAhmedMahmoud** [comment]
I am using swagger with NodeJS and  one of API response  size is `1.54MB` not working with me and also hanging

**Comment by blekinge** [comment]
I have at fix

Do something like this in your webservice method
```java
//https://github.com/swagger-api/swagger-ui/blob/master/src/core/components/response-body.jsx#L61
//If you set the Content-Disposition to something containing attachment, it will not be rendered but made downloadable in swagger UI
//inline will be rendered directly in browser, if invoked not through swagger UI
httpServletResponse.setHeader("Content-Disposition", "inline; swaggerDownload=\"attachment\"; filename=\"filename.txt\"");
```
I.e. set the Content-Disposition header to something containg attachment


If you look at <https://github.com/swagger-api/swagger-ui/blob/master/src/core/components/response-body.jsx#L61> you will see that swagger UI checks if the HTTP header "Content-Disposition" contains the string "attachment"
If it does, the content is NOT shown directly, but hidden behind a download link.
Note that response-body.jsx does NOT check anything about the well-formedness of the content disposition header. It just checks if it contains the word "attachment".
For this reason I just dumped some extra field `swaggerDownload="attachment";`, in order to have the word "attachment" there, without confusing other systems.

If this API method is invoked normally from a browser, rather than from swagger UI, it would just render the content inline, as this is specified, and also the default behaivour if Content-Disposition is not specified

**Comment by bovino** [comment]
2021 here, using swagger-ui-express with node ( https://www.npmjs.com/package/swagger-ui-express ) latest version (4.1.6) and this problem is still there :-D

**Comment by Kraego** [comment]
I have the same situation with ```asp .net 5``` (the repsonse body is large, but far from huge 650kB json)

```
    <PackageReference Include="Swashbuckle.AspNetCore" Version="5.6.3" />
    <PackageReference Include="Swashbuckle.AspNetCore.Annotations" Version="5.6.3" />
    <PackageReference Include="Swashbuckle.AspNetCore.Newtonsoft" Version="5.6.3" />
    <PackageReference Include="Swashbuckle.AspNetCore.Swagger" Version="5.6.3" />
    <PackageReference Include="Swashbuckle.AspNetCore.SwaggerUI" Version="5.6.3" />
```

**Comment by BjoernHund** [comment]
We have the same problem. The swagger ui does not respond.

```
    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.1.4" />
    <PackageReference Include="Swashbuckle.AspNetCore.Annotations" Version="6.1.4" />
```

**Comment by BjoernHund** [comment]
> @Kraego, @BjoernHund as mentioned above, you can speedup rendering by disabling syntax highlight. Here is example for AspNetCore.
> 
> ```cs
>  app.UseSwaggerUI(config =>
>  {
>     config.ConfigObject.AdditionalItems["syntaxHighlight"] = new Dictionary<string, object>
>     {
>         ["activated"] = false
>     };
> });
> ```

Yes, this solution works, but that's not the solution for the future.

**Comment by Kraego** [comment]
> 
> 
> @Kraego, @BjoernHund as mentioned above, you can speedup rendering by disabling syntax highlight. Here is example for AspNetCore.
> 
> ```cs
>  app.UseSwaggerUI(config =>
>  {
>     config.ConfigObject.AdditionalItems["syntaxHighlight"] = new Dictionary<string, object>
>     {
>         ["activated"] = false
>     };
> });
> ```

Yes i can also confirm that it's a working workaround for the problem.  @Lex45x c thanks you.

**Comment by spyro2000** [comment]
5. Years. Later. Still the same problem and no flag available to disable pretty print... :(

**Comment by lamaan** [comment]
I may take a look at this, as the problem is an issue for me now. But for those who want to dig into this one, this will most likely be an issue with string concatenation in the body rendering. This usually involves the expensive process of allocating memory many times if done the wrong way (e.g. just using stringc = stringa + stringb repeatedly). I have seen this kind of issue several times in the past. Instead build a array of strings and then concatenate it.

**Comment by Omzig** [comment]
> I may take a look at this, as the problem is an issue for me now. But for those who want to dig into this one, this will most likely be an issue with string concatenation in the body rendering. This usually involves the expensive process of allocating memory many times if done the wrong way (e.g. just using stringc = stringa + stringb repeatedly). I have seen this kind of issue several times in the past. Instead build a array of strings and then concatenate it.

Yep, it used to not be a problem prior to v5.5.1

**Comment by jl398m** [comment]
The issue doesn't seem to be caused by response size, but the number of records. I have a result with only 2 columns (int, date) but result sets >1000 records cause the UI to be awful, even though the total response is < 1mb. The same call with Postman renders in about a second. 

Since swagger UI is only for demonstrating how to use the API, I'd like an easy way to set a max number of rendered records in swagger. Is that possible?

**Comment by jl398m** [comment]
And turning off syntax highlighting helped a lot, but I'd still rather not have swagger try to display so many records.

**Comment by taranlu-houzz** [comment]
I have recently hit this issue as well while working on an internal api using `fastapi` that can return `json` with many entries. I feel like the following combination of tweaks/options would be helpful:

- Render in two stages:
  - Unhighlighted first and run highlighting asynchronously
  - Swap in highlighted version when done
- Add a non-blocking icon/widget spinner associated with the response body that indicates its content is being worked on
- Option to cap max records with good default
- Option to cap max response length with good default

Unfortunately, I can't offer to help with this right now, and I do get that changes like this would be a fair amount of work. Just wanted to add my two cents and note that this is an issue I have also run into.

**Comment by masc3d** [comment]
>   * Unhighlighted first and run highlighting asynchronously

👍🏼 https://github.com/swagger-api/swagger-ui/issues/3832#issuecomment-674387896
high workloads do not belong on ui thread.

**Comment by taranlu-houzz** [comment]
@masc3d Yes, good point. It should either be a separate thread, or a subprocess that is run asynchronously.

**Comment by sebastianhaberey** [comment]
The workaround for those that use Spring (in application.yml):

```
springdoc:
  swagger-ui:
    syntax-highlight:
      activated: false
```

**Comment by spyro2000** [comment]
Wish this could be automated if payload > x MB or something.

**Comment by andiemmadavies** [comment]
The disabling syntax highlighting workaround didn't work for me, so I kept exploring. I couldn't possibly think what was causing it to hang so badly, as my controller method and associated data structures weren't particularly chunky. Then I realised that I'm declaring in an attribute that my method will be returning a `BadHttpRequestException` in cases where I'm doing validation on the request data:

![Code screenshot](https://user-images.githubusercontent.com/17612522/181016822-db98fd9d-f363-4e52-9065-a1f109a88429.png)

When seeing that the Swagger UI generates an example instance of this return type, this results in an enormous amount of JSON as Swagger populates every possible property recursively, which is what's causing the hanging in every single interaction in my case:

![Generated JSON](https://user-images.githubusercontent.com/17612522/181017778-10fea23c-464a-4abe-b0f2-f0549bf1511b.png)

Removing my usage of this attribute to declare this possible return type is what made the UI responsive for me again. This is obviously a workaround and not a solution, but I thought I'd share my findings here in case it helps other folk stop tearing their hair out!

**Comment by LouiseYau** [comment]
> SwaggerUI({
>         syntaxHighlight: {
>           activated: false,
>           theme: "agate"
>         },
>         //url: path,
>         ....
>       });

Is there anyway this workaround can be implemented in swagger-ui-react v3.0.6? syntaxHighlight isn't an exposed prop

**Comment by ad9999** [comment]
Any update / fix available for `swagge-ui-react` package ?

**Comment by JoelYoung01** [comment]
I've run into the same issue, Swagger will hang after receiving a response from my API, the response takes about 2 seconds but the swagger page will hang for a hot minute before actually showing the result. I would dive into this if I was more experienced, and I might still try, but another proposed fix that I already googled for would be to disable highlighting for specific endpoints in the docs for known issue endpoints. Not sure what is easier to implement.

**Comment by spyro2000** [comment]
Is this lib still actively maintained? This issue is now unfixed for several years :(

**Comment by heldersepu** [comment]
> Is this lib still actively maintained? This issue is now unfixed for several years :(

Yes, just check the releases page:
https://github.com/swagger-api/swagger-ui/releases

This issues is not high priority, most developers do not have API with large responses and when large data is needed pagination helps reduce the amount of data retrieved at once

**Comment by guillaume-alvarez** [comment]
> I've run into the same issue, Swagger will hang after receiving a response from my API, the response takes about 2 seconds but the swagger page will hang for a hot minute before actually showing the result. I would dive into this if I was more experienced, and I might still try, but another proposed fix that I already googled for would be to **disable highlighting for specific endpoints in the docs for known issue endpoints**. Not sure what is easier to implement.

That would indeed solve the issue in my case.

**Comment by mImranAziz** [comment]
2,3 megabytes are slowing down the browser. I don't think this is huge amount of data

**Comment by ocmob** [comment]
Setting `content-type` to `application/octet-stream`, as well as `content-disposition` to `attachment` will force the swagger to display "Download file" link instead of rendering the contents.

**Comment by kenjdavidson** [comment]
Seems to still be an issue when switching to `swaggerUI "org.webjars:swagger-ui:5.9.0"`, getting a response back where the content size is ~2MB and the error is:

```
react-dom.production.min.js:216 RangeError: Maximum call stack size exceeded
    at j (highlight.js:256:52)
    at M (highlight.js:412:16)
    at si (react-dom.production.min.js:157:137)
    at $i (react-dom.production.min.js:180:154)
    at Ks (react-dom.production.min.js:269:343)
    at Il (react-dom.production.min.js:250:347)
    at Ol (react-dom.production.min.js:250:278)
    at Cl (react-dom.production.min.js:250:138)
    at bl (react-dom.production.min.js:243:163)
    at react-dom.production.min.js:123:115
```

Turning `syntaxHighlight` off resolves the issue, but obviously makes it a little less nice to look at.   As mentioned above, not a huge issue, but just following up to provide some more context.

**Comment by krocofish** [comment]
When it will FIXED ? In 10 years ?

**Comment by spyro2000** [comment]
Well, it's already been 7 years, so...

**Comment by heldersepu** [comment]
... my 2cents

Remember this is a free and open source project, the fastest way to get something done is to do it yourself ... that is what I did when I was using swagger often, if there was something that really bothered me I would spend the time to get familiar with the code and send as PR with a fix or workaround

Now specifically to this issue, I run into this only in theory, I never had a production API with this issue any large responses should have pagination to reduce exactly this issue as other consumers will run into this same issue

**Comment by joebowbeer** [comment]
@heldersepu wrote:
> Now specifically to this issue, I run into this only in theory, I never had a production API with this issue any large responses should have pagination to reduce exactly this issue as other consumers will run into this same issue

I suspect that most of the people who have been watching and upvoting this are doing so because they are experiencing practical not theoretical problems.

**Comment by heldersepu** [comment]
> I suspect that most of the people who have been watching and upvoting this are doing so because they are experiencing practical not theoretical problems.

Yes I'm sure they do... I would recommend as the best fix refactor the API to return small responses using pagination

**Comment by plachor** [comment]
There are endpoints where pagination make sense and where it is not an option. You may need to expose large object which cannot be returned in parts but needs to be exposed in single request.

Adding feature killing browser whenever you execute such endpoint is bad design. 

Same for possible responses. I hate my API that it is telling in schema what are the default error responses like not authorized or forbidden since when you expand any endpoint you need to scroll for hours to find next endpoint. 

Please do not explain bad implementation with excuses.

**Comment by heldersepu** [comment]
@plachor Can you provide an example of a case where pagination is not an option, I honestly have not come across it.

In the past few years everyone I talk to is moving toward a microservices architecture, smaller parts and browser makes many calls retrieving partial data... think google maps, instead of loading a huge image we have small map tiles and the browsers makes many requests and then combines them in a nice map

**Comment by heldersepu** [comment]
... to all that contributed improving the performance of Swagger-UI with large response, awesome work, we appreciate your time. 
the initial report was for:
 https://petstore.swagger.io/#/pet/findPetsByStatus
with latest version that now loads really fast, no longer an issue

... to all that are still running into this issue consider contributing to this project: 
- very least provide a link to the API that reproduces the problem, 
- provide some ideas of how to fix it
- ideally spend some time troubleshooting and debugging the code

**Comment by jmsteinbrunner** [comment]
Alternatively, if you don't have time to contribute but want this to work with large payloads, simply use the workaround posted in the previous [comment](https://github.com/swagger-api/swagger-ui/issues/3832#issuecomment-698531637) from September 2000.  I continue to use this workaround for older APIs I don't have time to rework and paginate.

**Comment by krocofish** [comment]
my 5 cents - About pagination ... Just imagine - our frontend is like excel app and works with data stored in mongodb . All data are json and User open page he need all data . Do we need pagination ? I don't think so ... By the way json data with 1000 rows weight 1 mb , we are using socket .. But for our purpose we have swagger on back and it crashed sometimes - ok - no probs - we started using other tools . No probs 

![image](https://github.com/swagger-api/swagger-ui/assets/139951944/f7bd55ce-d67e-4b7e-b245-d0f2bef9e09c)

**Comment by heldersepu** [comment]
> my 5 cents - About pagination ... Just imagine - our frontend is like excel app and works with data stored in mongodb . All data are json and User open page he need all data . Do we need pagination ? I don't think so ... By the way json data with 1000 rows weight 1 mb , we are using socket .. But for our purpose we have swagger on back and it crashed sometimes - ok - no probs - we started using other tools . No probs
> 
> 

Yes I would 100% add pagination to that, that is the perfect case! 
For a nice user experience when the page first load user only needs to see the top 20 some rows, of course show a loading animation on the corner while the rest of the data loads async in multiple batches or just  load the rest of the data on a specific user event like applying a filter or scrolling down.

...and yes please do use other tools, I mentioned that a while back:
https://github.com/swagger-api/swagger-ui/issues/3832#issuecomment-674574538 
> ... I would not use a browser to look at large data, instead, I get the schema definition, and load the data into a database for further processing/analysis.

**Comment by heldersepu** [comment]
> Alternatively, if you don't have time to contribute but want this to work with large payloads, simply use the workaround posted in the previous [comment](https://github.com/swagger-api/swagger-ui/issues/3832#issuecomment-698531637) from September 2000. I continue to use this workaround for older APIs I don't have time to rework and paginate.

100x :+1:

Another alternative for those that want to see this resolved and don't like any of the existing workarounds, also don't want to spend the time troubleshooting and debugging this, consider hiring someone

**Comment by plachor** [comment]
> @plachor Can you provide an example of a case where pagination is not an option, I honestly have not come across it.
> 
> In the past few years everyone I talk to is moving toward a microservices architecture, smaller parts and browser makes many calls retrieving partial data... think google maps, instead of loading a huge image we have small map tiles and the browsers makes many requests and then combines them in a nice map

Let's try. 

In one of my cases I expose Elasticsearch index mapping and configuration, another endpoint exposes node configuration. If I try to run them browser hangs for a minute.

So each time you have a large object to return which is needed complete to start analysis. Case in which paging would only add spare code on both sides. Or would more look like chunking.

There can be endpoint returning JSON documents created by end user in some envelope.

**Comment by taranlu-houzz** [comment]
Just to add to this: not all apps that use Swagger are production apps that have many users. I have have some internal tools that use FastAPI (which uses Swagger), and it is more convenient to return large json objects directly in some cases.

**Comment by heldersepu** [comment]
@Plachor & @taranlu-houzz  ...You might have missed my previous comment:

... consider contributing to this project:
- very least provide a link to the API that reproduces the problem,
- provide some ideas of how to fix it
- ideally spend some time troubleshooting and debugging the code

You don't like any of the existing workarounds, also don't want to spend the time troubleshooting and debugging this or don't know how to, consider hiring someone that can
shoutout to @kenjdavidson I'm sure he can fix this for you, go sponsor him:
https://github.com/sponsors/kenjdavidson

**Comment by spyro2000** [comment]
... or you just use a better JSON printer like https://chromewebstore.google.com/detail/json-formatter/bcjindcccaagfpapjjmafapmmgkkhgoa?pli=1

This pretty-prints even super-large JSON in 2 seconds while the same JSON kills Swagger for 10s of minutes.

Or, please, just deactivate pretty-printing at all if > 1MB or so, please.

Please.

**Comment by heldersepu** [comment]
@spyro2000 That is a Chrome extension I don't think that can be integrated into Swagger-UI... 
but go ahead give it try, let us know how it goes

**Comment by heldersepu** [comment]
... and if any one wants to implement the idea to deactivate pretty-printing if response is greater than X, start here:
https://github.com/search?q=repo%3Aswagger-api%2Fswagger-ui%20canSyntaxHighlight&type=code
The `canSyntaxHighlight` is the variable that will turn on/off that

**Comment by kristiansamuelsson** [comment]
Thanks for the hint on where to start, @heldersepu!
I have implemented two new config options to allow for:
- disabling rendering if payload size is over a specified number of bytes, and/or
- disabling pretty-printing if a payload size is over a specified number of bytes

**Comment by heldersepu** [comment]
@kristiansamuelsson  way to go!
Looking at your profile that is your first PR on GitHub, congrats! ... 
Do you mind sharing your experience, 
- how hard was to debug/code that?
- how long did it take you?  

Maybe that inspires others

**Comment by glowcloud** [comment]
Hi everyone, just letting you know I’m investigating this issue and testing [this PR](https://github.com/swagger-api/swagger-ui/pull/9625) proposing a fix with a specification from [this issue](https://github.com/swagger-api/swagger-ui/issues/9662).

Unfortunately, it seems like only adding this render limit is not enough. You can check it out with these steps (using https://github.com/swagger-api/swagger-ui/pull/9625):
1. Load the [mentioned spec](https://github.com/swagger-api/swagger-ui/files/14486492/openapi.yaml.zip) while setting a low render limit
2. Navigate to `CRUD - AGM Disease Annotations`
3. Click on the first PUT request
4. Wait for it to load
5. Click on try it out
6. Wait for the site to crash 

I also tried adding the limit to other components, ex. `RequestBodyEditor`, `Curl`, but it hasn't fixed the crash when going through the mentioned steps.

The memory usage is still high when just expanding the operation, and even when disabling rendering for some of the components (`RequestBody`, `Response`). I think that the issue might partially lie with immutable types, at least from checking the memory usage. I will be investigating this further.

**Comment by heldersepu** [comment]
@glowcloud I'm not sure I follow ... those steps have nothing to do with this issue and large response bodies

**Comment by nojaf** [comment]
Hello,

We are also experiencing this issue. After investigating, I've found that `react-syntax-highlighter` seems to be the primary cause of the bottleneck:

![Screenshot 2024-03-20 110733](https://github.com/swagger-api/swagger-ui/assets/2621499/ca65b7ad-fc11-40ef-99b5-2c03534924c1)

Is there any possibility of replacing this component? I conducted a preliminary test by swapping it out for [`@monaco-editor/react`](https://www.npmjs.com/package/@monaco-editor/react) and immediately observed significantly improved performance with large results.

![Screenshot 2024-03-20 113832](https://github.com/swagger-api/swagger-ui/assets/2621499/4cc87077-3352-49fa-a96d-2c6a5e4a52d9)

I'm ready to submit a pull request if the maintainers are open to this change.

**Comment by heldersepu** [comment]
@nojaf You don't have to ask permission to submit a PR ... 
now regarding [@monaco-editor/react](https://www.npmjs.com/package/@monaco-editor/react) that seems to be a full editor not just syntax highlighter, maybe there are some feature that can be turned off

**Comment by nojaf** [comment]
> You don't have to ask permission to submit a PR ...

No, it makes more sense to find out if the maintainers would ever consider merging it before I spent my time creating the PR.

The Monaco editor can be tweaked to appear very minimal (readonly mode, remove line numbers, side bar, ...)

**Comment by heldersepu** [comment]
@nojaf no worries project still active, in my experience any PR fixes a known issue will be accepted...
it is great that The Monaco editor can be tweaked to appear very minimal, but my concern is size: 

https://www.npmjs.com/package/@monaco-editor/react
Unpacked Size
152 kB. <- that is great if is truly only that

https://www.npmjs.com/package/monaco-editor
Unpacked Size
92.2 MB <- now this is on the large side

From your preliminary work looks you have it working... in your environment how much bugger did the `bundle` get  https://github.com/swagger-api/swagger-ui/tree/master/dist ?

**Comment by heldersepu** [comment]
@nojaf that is great, a final bundle of `1.3 MiB` is no significant change in size

**Comment by spyro2000** [comment]
Pah, just leave it broken and crashing for another few years and enjoy the 150 kB saved each time. Priorities!

**Comment by nojaf** [comment]
Ok @heldersepu, I went ahead and raised https://github.com/swagger-api/swagger-ui/pull/9728.
There are definitely some open questions left. I appreciate your input and am willing to tweak things accordingly.

**Comment by heldersepu** [comment]
Anyone interested in testing the alpha prototype from @nojaf go to:
https://raw.githack.com/nojaf/swagger-ui/large-bodies/dist/index.html#/

This is your opportunity to influence how that will look like

**Comment by syednomanforyou** [comment]
app.UseSwaggerUI(config =>
{
    config.ConfigObject.AdditionalItems["syntaxHighlight"] = new Dictionary<string, object>
    {
        ["activated"] = false
    };
});

this improve performance

**Comment by char0n** [comment]
Hi everybody,

Let me first address the current state of this issue. Where we are and where we want to get.

Large response bodies are indeed causing issues. The issue with large response bodies starts with Immutable serialization into redux store and continues with rendering with or without syntax highlighting. 


We will address this problems progressively in multiple steps.

### Step 1

Give ability to disable syntax highlighting. As mentioned multiple times in this issue, syntax highlighting can be disabled fully by utilizing `syntaxHighlight.activated=false` option. This is already fully supported.

### Step 2

Detect possible large response bodies and disable syntax highlighting for them even when `syntaxHighlight.activated=true`. There is a PR that actually comes pretty close to our idea: https://github.com/swagger-api/swagger-ui/pull/9625

### Step 3

Detect possible large response bodies and disable rendering for them, but provide ability to download. This way we'll avoid any possible operation performed on large respose bodies expect storing it in redux store. We still have to provide ability to download the large response body though. Again https://github.com/swagger-api/swagger-ui/pull/9625 comes pretty close to our idea.

---

Entire syntax highlighting feature has been consolidated into single `syntax-highlighting` plugin in https://github.com/swagger-api/swagger-ui/pull/9783. This makes it possible to completely override how syntax highlighting is done or switch the syntax highlighter entirely. We opted to use `highlight.js` for default syntax highlighter as it's more performant than `prism`.

---

@nojaf,

Thank you for your [PR](https://github.com/swagger-api/swagger-ui/pull/9728), be we will not be incorporating your changes into the core SwaggerUI. We already use custom build MonacoEditor in [SwaggerEditor@5](https://github.com/swagger-api/swagger-editor/tree/next). We have to be very careful what dependencies we pull into our bundles and how they affect each other. Using MonacoEditor is a major arch decision, we'll rather provide progressive enhancement to this issue as described in previous steps. 

The https://github.com/swagger-api/swagger-ui/pull/9783 opened possibility to utilize custom syntax highlighting mechanism, and I've seen you've already transformed your PR into the plugin that other people can utilize if they choose to. 

---

@heldersepu,

> https://petstore.swagger.io/?docExpansion=None&defaultModelsExpandDepth=None&url=https://gist.githubusercontent.com/heldersepu/ab37902d2f687d2c71aa1b36bb72ee48/raw/0f68906074dfbad5528a81132ae7975b6bf2ef75/openapi.yml#/CRUD%20-%20AGM%20Disease%20Annotations

This particular definition according to my research, suffers with different issues not related to large response bodies. When expanding operation most time is spent in `dereferencing` (resolution) and JSON Schema sample generation. Only after that (when browser still hasn't crashed) the syntax highliting issue comes into light.

**Comment by nicoloschimizzi** [comment]
Hi, can ask how it is possible to set ```syntaxHighlight.activated=false``` having Swagger in a Docker system? Thank you in advance.

**Comment by jcjolley** [comment]
Is this still in progress?  We have several endpoints that stream a large amount of data that developers from other teams want to test in swagger before setting up a rest client in their apps to hit our endpoints, but they've lost confidence in our application because they interpret Swagger hanging as our application hanging.  I've tried directing them to curl the endpoints to see the true performance but it's not as convenient as swagger and they don't do it.

**Comment by spyro2000** [comment]
AFAIK this has never been fixed. At least not internally. :(

**Comment by heldersepu** [comment]
@jcjolley do you have a sample endpoint of your api we can use to test?
for most disabling syntaxHighlight is enough
